great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -293,3 +359,23 @@ def _determine_substitutions(self) -> dict:
         }
 
         return substitutions
+
+    def get_config_with_variables_substituted(
+        self, config: Optional[DataContextConfig] = None
+    ) -> DataContextConfig:
+        """
+        Substitute vars in config of form ${var} or $(var) with values found in the following places,
+        in order of precedence: ge_cloud_config (for Data Contexts in GE Cloud mode), runtime_environment,
+        environment variables, config_variables, or ge_cloud_config_variable_defaults (allows certain variables to
+        be optional in GE Cloud mode).
+        """
```

Comment:
While it isn't particularly detrimental here, I think the less we reference child classes in `AbstractDataContext`, the better. At this point in the inheritance hierarchy, we have no knowledge of things like `CloudDataContext`.

I think as much as possible, we should treat the methods herein as black boxes and have the child classes fill in the details. The order of precedence is determined by the child class and can be abstracted away by the `_determine_substitutions` method from the child. Perhaps this documentation better fits there?

All that being said, I know this refactor is less than straightforward 😄 just want to try and keep this class as abstract as possible.

Reply:
This is actually being addressed in #5339 where we are intending to get rid of the hierarchy and have usage_stats be disabled if an override is set in any of the above-mentioned locations . I'd like to propose that we maintain this functionality for this PR to preserve current behavior (#5339 will require modifying behavior and some related tests)